application-analytics-ui/src/main/resources/Analytics/Code/Macros/VisitsSummary.xml

```diff
@@ -0,0 +1,277 @@
+    editCanvas(canvas, cacheId, timeInterval, endpoints,fields ,period, yAxeNames);
+  });
+
+  // Function to create a new chart.
+  function createChart(canvas, labels, values, yAxeNames) {
+    let data = {
+      labels: labels,
+      datasets: []
+    };
+    yAxeNames.forEach((yAxeName, i) =&gt; {
+      data.datasets.push({
+        label: yAxeName,
+        borderColor: RGB[i],
+        data: values[i],
+        pointRadius: 0,
+        pointHoverRadius: 5,
+        // Makes the second line wider to be able to view the overlapping parts
+        borderWidth: 2 + i*3+1
+      });
+    });
+    return new Chart(canvas, {
+      type: 'line',
+      data: data,
+      options: {
+        scales: {
+          x: {
+            ticks: {
+              autoSkip: true,
+              maxRotation: 90,
+              autoSkipPadding: 3,
+              maxTicksLimit: 40
+            }
+          },
+          y: {
+            beginAtZero: true
+          }
+        },
+        plugins: {
+          tooltip: {
+            mode: 'index',
+            intersect: false,
+          }
+        },
+        interaction: {
+          mode: 'nearest',
+          intersect: true
+        }
+      }
+    });
+  }
+  //  Makes a request to an endpoint and returns the promise.
+  function makeRequest(url, queryString) {
+    return new Promise((resolve, reject) =&gt; {
+      $.getJSON(`${url}?${queryString}`)
+        .done(resolve)
+        .fail(function (jqxhr, textStatus, error) {
+          const err = `${textStatus}, ${error}`;
+          console.error(`Request Failed: ${err}`);
+          reject(err);
+        });
+    });
+  }
+
+  // Updates the chart.
+  function editCanvas(canvas, cacheId, date, endpoints, fields, period, yAxeNames) {
+    let requestPageRef = XWiki.Model.resolve('Analytics.Code.Macros.VisitsSummaryJson', XWiki.EntityType.DOCUMENT);
+    let url = new XWiki.Document(requestPageRef).getURL('get');
+    let requests = []; // This is correctly defined
+    // A field can have multiple endpoints(lines) call all of them and store the promise in vector for later use.
+    for (let i = 0; i &lt; endpoints.length; i++) {
+      let queryString = $.param({
+        'date': date,
+        'apiMethod': endpoints[i],
+        'period': period
+      });
+      requests.push(makeRequest(url, queryString));
+    }
+    // After all the promises are completed update the chart with the new data.
+    Promise.all(requests).then(responses =&gt; {
+      let chart = chartsMap.get(cacheId);
+      if (chart != undefined) {
+        chart.destroy();
+      }
+      let values = fields.map((field, i) =&gt; responses[i][field]);
+      // Map the chart to the id of the macro to be able to edit the canvas.
+      chartsMap.set(cacheId, createChart(canvas, responses[0]["labels"], values, yAxeNames));
+      canvas.removeClass("loading");
+    }).catch(error =&gt; {
+      console.error("Request failed while in the VisitsSummary macros:", error)
```

Comment:
You could display a notification, so the user knows something went wrong https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/FrontendResources/XWikiNotifications/
You could keep the console log for a detailed message with the failure reason, which might be needed for debugging purposes